File: src/chunker.py
import re

from .config import MAX_CHUNK_CHARS, OVERLAP_CHARS, TARGET_CHUNK_CHARS
from .markdown_parser import clean_text, split_into_blocks


def split_large_block(block: str, max_chars: int) -> list[str]:
    """Split an oversized block at sentence boundaries where possible."""
    if len(block) <= max_chars:
        return [block]

    sentences = re.split(r"(?<=[.!?])\s+", block)
    pieces = []
    current = ""

    for sentence in sentences:
        sentence = sentence.strip()
        if not sentence:
            continue

        if len(current) + len(sentence) + 1 <= max_chars:
            current = f"{current} {sentence}".strip()
            continue

        if current:
            pieces.append(current)

        if len(sentence) > max_chars:
            for start in range(0, len(sentence), max_chars):
                pieces.append(sentence[start:start + max_chars].strip())
            current = ""
        else:
            current = sentence

    if current:
        pieces.append(current)
    return pieces


def get_overlap(text: str, overlap_chars: int) -> str:
    """Return the end of a chunk, starting at a sentence boundary if possible."""
    if len(text) <= overlap_chars:
        return text

    overlap = text[-overlap_chars:]
    match = re.search(r"[.!?]\s+", overlap)
    if match:
        overlap = overlap[match.end():]
    return overlap.strip()


def build_chunk(text: str, section: dict) -> dict:
    """Build a chunk with section metadata and basic statistics."""
    text = clean_text(text)
    return {
        "chunk_id": None,
        "chunk_index": None,
        "section": section["heading"],
        "section_level": section["heading_level"],
        "section_id": section["section_id"],
        "text": text,
        "char_count": len(text),
        "word_count": len(text.split()),
    }


def create_chunks(sections: list[dict]) -> list[dict]:
    """Convert sections into meaningful, overlapping chunks."""
    chunks = []

    for section in sections:
        processed_blocks = []
        for block in split_into_blocks(section["content"]):
            if len(block) > MAX_CHUNK_CHARS:
                processed_blocks.extend(split_large_block(block, MAX_CHUNK_CHARS))
            else:
                processed_blocks.append(block)

        current_blocks = []
        current_length = 0
        for block in processed_blocks:
            block_length = len(block)
            if current_blocks and current_length + block_length > TARGET_CHUNK_CHARS:
                chunk_text = "\n\n".join(current_blocks)
                chunks.append(build_chunk(chunk_text, section))
                overlap_text = get_overlap(chunk_text, OVERLAP_CHARS)
                current_blocks = [overlap_text]
                current_length = len(overlap_text)

            current_blocks.append(block)
            current_length += block_length + 2

        if current_blocks:
            chunks.append(build_chunk("\n\n".join(current_blocks), section))

    for index, chunk in enumerate(chunks):
        chunk["chunk_id"] = f"chunk_{index + 1:04d}"
        chunk["chunk_index"] = index

    return chunks

File: web/lib/src/chunker.ts
import {
  MAX_CHUNK_CHARS,
  OVERLAP_CHARS,
  TARGET_CHUNK_CHARS,
} from "./config";

import {
  cleanText,
  splitIntoBlocks,
} from "./markdown_parser";

import type { Section, Chunk } from "./types";


export function splitLargeBlock(
  block: string,
  maxChars: number
): string[] {
  if (block.length <= maxChars) {
    return [block];
  }

  const sentences = block.split(/(?<=[.!?])\s+/);

  const pieces: string[] = [];
  let current = "";

  for (const sentence of sentences) {
    const trimmedSentence = sentence.trim();

    if (!trimmedSentence) {
      continue;
    }

    if (
      current.length +
        trimmedSentence.length +
        1 <=
      maxChars
    ) {
      current = `${current} ${trimmedSentence}`.trim();
      continue;
    }

    if (current) {
      pieces.push(current);
    }

    if (trimmedSentence.length > maxChars) {
      for (
        let start = 0;
        start < trimmedSentence.length;
        start += maxChars
      ) {
        pieces.push(
          trimmedSentence
            .slice(start, start + maxChars)
            .trim()
        );
      }

      current = "";
    } else {
      current = trimmedSentence;
    }
  }

  if (current) {
    pieces.push(current);
  }

  return pieces;
}

export function getOverlap(
  text: string,
  overlapChars: number
): string {
  if (text.length <= overlapChars) {
    return text;
  }

  let overlap = text.slice(-overlapChars);

  const match = overlap.search(/[.!?]\s+/);

  if (match !== -1) {
    const boundary = overlap.match(/[.!?]\s+/);

    if (boundary && boundary.index !== undefined) {
      overlap = overlap.slice(
        boundary.index + boundary[0].length
      );
    }
  }

  return overlap.trim();
}

export function buildChunk(
  text: string,
  section: Section
): Chunk {
  const cleanedText = cleanText(text);

  return {
    chunkId: null,
    chunkIndex: null,
    section: section.heading,
    sectionLevel: section.headingLevel,
    sectionId: section.sectionId,
    text: cleanedText,
    charCount: cleanedText.length,
    wordCount: cleanedText.split(/\s+/).filter(Boolean).length,
    prevText: "",
  };
}

export function createChunks(
  sections: Section[]
): Chunk[] {
  const chunks: Chunk[] = [];

  for (const section of sections) {
    const processedBlocks: string[] = [];

    for (const block of splitIntoBlocks(section.content)) {
      if (block.length > MAX_CHUNK_CHARS) {
        processedBlocks.push(
          ...splitLargeBlock(
            block,
            MAX_CHUNK_CHARS
          )
        );
      } else {
        processedBlocks.push(block);
      }
    }

    const currentBlocks: string[] = [];
    let currentLength = 0;

    for (const block of processedBlocks) {
      const blockLength = block.length;

      if (
        currentBlocks.length > 0 &&
        currentLength + blockLength >
          TARGET_CHUNK_CHARS
      ) {
        const chunkText =
          currentBlocks.join("\n\n");

        chunks.push(
          buildChunk(chunkText, section)
        );

        currentBlocks.length = 0;
        currentLength = 0;
      }

      currentBlocks.push(block);

      currentLength += blockLength + 2;
    }

    if (currentBlocks.length > 0) {
      chunks.push(
        buildChunk(
          currentBlocks.join("\n\n"),
          section
        )
      );
    }
  }

  chunks.forEach((chunk, index) => {
    chunk.chunkId = `chunk_${String(
      index + 1
    ).padStart(4, "0")}`;

    chunk.chunkIndex = index;
  });

  return chunks;
}


export function appendPreviousText(
  chunks: Chunk[]
): Chunk[] {
  let previousText = "";

  for (const chunk of chunks) {
    chunk.prevText = previousText;

    previousText = getOverlap(
      String(chunk.text),
      OVERLAP_CHARS
    );
  }

  return chunks;
}